vacationtime-solution/src/Code.js

```diff
@@ -0,0 +1,195 @@
+  return notifiedColRange;
+}
+
+/**
+ * Creates a calendar event for an employee whose vacation
+ * request has been approved.
+ * @param {String} managerEmail Manager's email.
+ * @param {String} employeeName Name of employee.
+ * @param {String} employeeEmail Email of employee.
+ * @param {date} startDate Vacation request start date.
+ * @param {date} endDate Vacation request end date.
+**/
+function createCalEvent(managerEmail, employeeName, employeeEmail,
+                       startDate, endDate) { 
+    var managerCal = CalendarApp.getCalendarById(managerEmail);
+    // Creates a calendar event.
+    var event = managerCal.createEvent('APPROVED VACATION TIME FOR ' +
+        employeeName, startDate, endDate, {
+        description: 'Your vacation time from ' +
+                                            startDate + ' to ' + endDate +
+                                            ' has been approved! Enjoy!',
+        guests: employeeEmail,
+        sendInvites: true,
+    });
+}
+
+/**
+ * Sends e mails to employees whose vacation time request
+ * was NOT approved.
+ * @param {String} employeeEmail Email of employee.
+**/
+function sendEmails(employeeEmail) {
+  MailApp.sendEmail(employeeEmail, REJECTION_EMAIL_SUBJECT, 
+                        REJECTION_EMAIL_MESSAGE);
+}
+
+/**
+ * Checks the approval status of each employee and notifies
+ * them of their status accordingly.
+ * @param {String} employeeEmail Email of employee.
+ * @param {String} managerEmail Manager's email.
+ * @param {String} employeeName Name of employee.
+ * @param {String} approvalStatus Manager-set status.
+ * @param {date} startDate Vacation request start date.
+ * @param {date} endDate Vacation request end date.
+**/
+function checkApproval(employeeEmail, managerEmail, employeeName, 
+                       approvalStatus, startDate, endDate) {
+    var sheet = SpreadsheetApp.getActiveSheet();
+    var managerCal = CalendarApp.getCalendarById(managerEmail);
+    
+    // Checks approval status.
+    if (approvalStatus == 'NOT APPROVED') {
+      // Sends email of disapproval.
+      sendEmails(employeeEmail);
+    } else if (approvalStatus == 'APPROVED') {
+      // Creates calendar event.
+      createCalEvent(managerEmail, employeeName, employeeEmail,
+                      startDate, endDate);
+    }
+}
+
+/**
+ * Checks the notification status of each employee and, if not notified,
+ * notifies them of their status accordingly, through use of helper 
+ * functions.
+ */
+function notifyEmployees() {
+  var sheet = SpreadsheetApp.getActiveSheet();
+  var lastRow = sheet.getLastRow();
+  var lastCol = sheet.getLastColumn();
+  var frozenRows = sheet.getFrozenRows();
+  var startRow = frozenRows + 1;
+  var numRows = lastRow - frozenRows;
+
+  // Gets  notified column values.
+  var notifiedColRange = createNotifiedColumn();
+  var notifiedStatus = notifiedColRange.getValues();
+
+  for (var i = 0; i <= lastRow - startRow; i++) {
```

Comment:
Instead of adding/subtracting startRow, why not just set i to startRow?

Reply:
i is used to access the array values which are zero-indexed!

Reply:
I'm not sure the loop logic is correct. Have you tested this with multiple rows?

The questionable part is how you're using `i` with rangeValues. It looks like you're setting rangeValues to the current row you're trying to read (currentStartRow = i + startRow). And you're setting the numRows parameter to 1 when doing so. So rangeValues *should* be length of 1.

But then you're using `i` again as the index. This works if there's only 1 row in the sheet since the loop only runs once with i = 0. If there are *two* rows, then the second row, if I'm following the code correctly, will fail since i = 1 and that should be out of range for the array.

So that gets back to the original comment -- if you're reading one row at a time like this then you don't need i to be 0 based as the you'll *always* use 0 as the first index on the range.

The alternate approach is to read the full range up front *before* the loop. Then when you iterate you can safely do things 0-based.

Reply:
I had problems with this code when I ran it with multiple rows - but my new changes now work with multiple rows. I just pushed the new code a bit ago (which is working great!) but let me know if there still should be any changes in there. Thanks so much Steve - lots of mindless debugging this morning when I should've just pinged you instead!!

Reply:
Looks much better now.

You could also collapse the two ranges now that you've moved creation of the column -- notifiedColumnRange and range both are the same set of rows.

Reply:
Since I am only calling createNotifiedColumn once, should I store notifiedColumnRange in a global variable so that I can access it every time I am running "notify employees"?

Reply:
No need to. Since it's just an added column, just expand the # of columns in getRange() to include it.